File: mtgSearch/src/androidTest/java/com/dbottillo/mtgsearchfree/database/SetDataSourceTest.java
package com.dbottillo.mtgsearchfree.database;

import android.database.Cursor;
import android.support.test.runner.AndroidJUnit4;

import com.dbottillo.mtgsearchfree.resources.MTGSet;

import org.junit.Test;
import org.junit.runner.RunWith;

import static junit.framework.Assert.assertNotNull;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

@RunWith(AndroidJUnit4.class)
public class SetDataSourceTest extends BaseDatabaseTest {

    @Test
    public void test_generate_table_is_correct() {
        String query = SetDataSource.generateCreateTable();
        assertNotNull(query);
        assertThat(query, is("CREATE TABLE IF NOT EXISTS MTGSet (_id INTEGER PRIMARY KEY, name TEXT,code TEXT)"));
    }

    @Test
    public void test_set_can_be_saved_in_database() {
        MTGSet set = new MTGSet(50);
        set.setName("Commander");
        set.setCode("CMX");
        long id = SetDataSource.saveSet(dataHelper.getWritableDatabase(), set);
        Cursor cursor = dataHelper.getReadableDatabase().rawQuery("select * from " + SetDataSource.TABLE + " where rowid =?", new String[]{id + ""});
        assertNotNull(cursor);
        assertThat(cursor.getCount(), is(1));
        cursor.moveToFirst();
        MTGSet setFromDb = SetDataSource.fromCursor(cursor);
        assertNotNull(setFromDb);
        assertThat(setFromDb.getId(), is(set.getId()));
        assertThat(setFromDb.getName(), is(set.getName()));
        assertThat(setFromDb.getCode(), is(set.getCode()));
        cursor.close();
    }
/*
    @Test
    public void test_card_are_unique_in_database() {
        SQLiteDatabase db = dataHelper.getWritableDatabase();
        int uniqueId = 444;
        MTGCard card = new MTGCard();
        card.setId(uniqueId);
        card.setCardName("one");
        long id = CardDataSource.saveCard(db, card);
        MTGCard card2 = new MTGCard();
        card2.setId(uniqueId);
        card2.setCardName("two");
        long id2 = CardDataSource.saveCard(db, card2);
        assertNotSame(id, id2);
        Cursor cursor = db.rawQuery("select * from " + CardDataSource.TABLE + " where _id =?", new String[]{uniqueId + ""});
        assertNotNull(cursor);
        assertThat(cursor.getCount(), is(1));
        cursor.moveToFirst();
        MTGCard cardFromDb = CardDataSource.fromCursor(cursor);
        assertThat(cardFromDb.getName(), is(card.getName())); // same id are ignored
        cursor.close();
    }

    @Test
    public void test_cards_can_be_retrieved_from_database() {
        MTGCard card = new MTGCard();
        card.setId(101);
        CardDataSource.saveCard(dataHelper.getWritableDatabase(), card);
        MTGCard card2 = new MTGCard();
        card2.setId(102);
        CardDataSource.saveCard(dataHelper.getWritableDatabase(), card2);
        Cursor cursor = dataHelper.getReadableDatabase().rawQuery("select * from " + CardDataSource.TABLE, null);
        ArrayList<MTGCard> cards = new ArrayList<>(cursor.getCount());
        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast()) {
                cards.add(CardDataSource.fromCursor(cursor));
                cursor.moveToNext();
            }
        }
        cursor.close();
        assertNotNull(cards);
        assertThat(cards.size(), is(2));
        assertThat(cards.get(0).getId(), is(card.getId()));
        assertThat(cards.get(1).getId(), is(card2.getId()));
    }*/
}

File: app/src/test/kotlin/com/dbottillo/mtgsearchfree/database/SetDataSourceTest.kt
package com.dbottillo.mtgsearchfree.database

import com.dbottillo.mtgsearchfree.model.MTGSet
import com.dbottillo.mtgsearchfree.storage.SetDataSource
import com.dbottillo.mtgsearchfree.util.readSetListJSON
import com.google.common.truth.Truth.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment

@RunWith(RobolectricTestRunner::class)
class SetDataSourceTest {

    private lateinit var underTest: SetDataSource
    private lateinit var mtgDatabaseHelper: MTGDatabaseHelper

    @Before
    fun setup() {
        mtgDatabaseHelper = MTGDatabaseHelper(RuntimeEnvironment.application)
        underTest = SetDataSource(mtgDatabaseHelper.readableDatabase)
    }

    @After
    fun close_data_helper() {
        mtgDatabaseHelper.close()
    }

    @Test
    fun `should generate right create table`() {
        val query = SetDataSource.generateCreateTable()
        assertThat(query).isNotNull()
        assertThat(query).isEqualTo("CREATE TABLE IF NOT EXISTS MTGSet (_id INTEGER PRIMARY KEY, name TEXT,code TEXT,type TEXT)")
    }

    @Test
    fun `should save sets inside database`() {
        val set = MTGSet(id = 5000, name = "Commander", code = "CMX")
        val id = underTest.saveSet(set)
        val cursor = mtgDatabaseHelper.readableDatabase.rawQuery("select * from " + SetDataSource.TABLE + " where rowid =?", arrayOf(id.toString() + ""))
        assertThat(cursor).isNotNull()
        assertThat(cursor.count).isEqualTo(1)
        cursor.moveToFirst()
        val setFromDb = underTest.fromCursor(cursor)
        assertThat(setFromDb).isNotNull()
        assertThat(setFromDb.id).isEqualTo(set.id)
        assertThat(setFromDb.name).isEqualTo(set.name)
        assertThat(setFromDb.code).isEqualTo(set.code)
        cursor.close()
        // need to clear up the db:
        underTest.removeSet(id)
    }

    @Test
    fun `should retrieve sets from database`() {
        val sets = underTest.sets
        assertThat(sets).isNotNull()
        assertThat(sets.size).isEqualTo(NUMBER_OF_SET) // the one added from the previous test
    }

    @Test
    fun `should load all sets correctly`() {
        val fromJson = readSetListJSON()
        val sets = underTest.sets
        assertThat(fromJson).isNotNull()
        assertThat(fromJson).containsAtLeastElementsIn(sets)
    }
}

private const val NUMBER_OF_SET = 233
